### DISCUSSION TITLE
Track Chunks and Merging

### DISCUSSION THREAD

**Discussion by lrpauley13** [melanchall/drywetmidi#307] [General]
Can you give me an example of merging track chunks into a Single track.  I have looked at your docs and can't quite figure it out. My end result is to take all tracks and merge them into one.  I don't know if merging to a single track is considered a midi 0 file.  I hope not because I would like to retain the lyrics.  I appreciate any help or just pointing me in the right direction.

**Comment by melanchall** [comment]
First of all, please say what kind of merging do you need: sequential (track chunks will go each after other) or parallel (all track chunks data will be played simultaneously)?

If you need parallel merging, then with this statement

> I would like to retain the lyrics

you need to be aware that:

1. lyrics will be retained, merging is not about throwing data away, it's just data reorganization;
2. you won't be able to say what notes lyrics corresponds in the result file because it all will be in a single track chunk.

**Reply by lrpauley13** [reply]
This might get a little long, but here goes. 

I am a retired programmer that always thought I would like to play the keyboard once I retire, which means I cannot read or play music.  I can still hear and see and sing, so I joined my church choir a number of years ago and enjoy the wonderful music we perform.  I struggle with some pieces because I cannot read music and do not hear as well as I did in my youth.

I can take a piece of sheet music that I am having problems learning and scan it and digitize it and make my computer play it to learn my part using software called SmartScore 64 Pro.  The software can create midi 0 and midi 1 files.

For a few years now, I have heard friends, relatives and other retirees say they would like to learn how to play.  I have looked at the little games available, but thought those do not help one learn sheet music and playing.  So, I thought why not look for some libraries on GitHub that would help me read and play a midi keyboard.  It would also keep my mind a little sharp using my retired skills as a developer.

My other thought was there is not enough music in the world.  Can I do something to change that just a little.  And I feel that is the exact reason you created your library, which is genius. 

In my search, I found 2 libraries.  One of course, your DryWetMidi, which is a backend midi library processor. The other  was MidiSheetMusic front end, which reads and shows sheet music as it plays.  Neither are mine, but could I merge them to make a better learning experience for noobie keyboard learners, like me.

I have worked on the modifications for a while, and it works fine for me right now because I am a SUPER NOOBIE, but not for others that are more advanced. So what I found is it seems my code works ok for single track midi 0 files, but not so well for midi 1 files.

I am including the project in hopes that you might find it entertaining (and probably laughable), but also a contribution to make your library even more used than it is today.  

The code is written to make use of a lighted midi keyboard.  I have an inexpensive Casio LK-S250. I have tried to make an "Adaptive Speed" checkbox with works with simple songs, but not more sophisticated arrangements.  Also, I have hard coded some of the options in the "**MidiOptions.cs**", which I will remove before uploading a new branch to MidiSheetMusic.  Most of the modifications to use your library are in the "MidiPlayer.cs".  

A few other things to note:  I am not a C# developer.  I started using VB in 1991.  I found C# to much like JavaScript and did not use it so much in my career.  So, if you look at the code remember I am not a C# programmer and did not write most of this code.

My end goal is find the current, iteration of notes played by the computer and progress the sheet music when the player hits the right notes.  You will find the code I am struggling with in the "**MidiPlayer.cs**" module.  I am sure my Linq statements are wrong and need adjustment.  

I cannot express how wonderful it is to come across your library.  I want to be able to play sheet music (a little), but more importantly, I want more people to play sheet music. If I can help in that effort, then great. 

I am uploading the project and the 2 files I test with.  I am sure you have plenty of midi files, but as a reference point I am including the files.  You should be able to go to the "Release" folder and execute the EXE.

If you don't have time to look at a full VS solution, I understand.  But again, thanks for all you have done for the music community.
[MidiSheetMusic-2.4-win-src.zip](https://github.com/user-attachments/files/16749856/MidiSheetMusic-2.4-win-src.zip)
[Little Drummer Boy, The.zip](https://github.com/user-attachments/files/16749868/Little.Drummer.Boy.The.zip)

**Reply by melanchall** [reply]
Thank you for such kind words regarding DryWetMIDI. I appreciate it! And thank you for your efforts to make a contribution to the music.

Yes, you're right, my intention to make the library was to bring together worlds of music and programming. I'm a programmer, and made some music compositions in rock/metal style (you can find them on my site mentioned in the profile if you want), was a drummer for short period of my life. But for .NET there were no libraries to solve popular music tasks easily without need to implement all those algorithms manually. So DWM is aimed to fill the gap.

As for music, I wanted to be a part of that world, to make a contribution, yes. So I definitely understand you. I respect those people who want to build a better world, to direct efforts for peaceful things that make the life a little better and easier.

Well, as for your task. You need to do something like that:

```csharp
var singleTrackChunk = trackChunksCollection.Merge();
```

This will create Type 0 track chunk. As I said all lyrics will be mixed together which can be a problem. To handle lyrics properly you need to learn how to work with multiple track chunks. Type 0 nowadays creates more problems than profit. It was reasonable for the past century where each byte had pretty big cost.

**Reply by melanchall** [reply]
@lrpauley13 Any news?
